Disable PDBs for hub/proxy, add PDB for autohttps, and relocate config proxy.pdb to proxy.chp.pdb

consideRatio opened the pull request:
This PR follows #1937 and discussion in #1934. I suggest we continue the discussion about _what to do_ in #1934 unless it is about the lines of codes changed.

I've opened https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/1951 to represent the wish for better documentation about HA so that the we are more explicit with assumptions.

## PR summary

- [x] Disabling hub/proxy PDB by default.
  This is done to close #1934, and motivated by them being non-HA applications. For more in-depth discussion, see #1934.
- [x] Added autohttps PDB, also disabled by default.
  This is done for completeness as it would generally make sense to enable PDBs for proxy/hub/autohttps all at once if either was enabled.
- [x] Renamed pdb configuration proxy.pdb to to proxy.chp.pdb.
  This was done to make the configuration more cohesive and easy to understand now that we have one for the proxy pod and the autohttps pod. This is in alignment of how we have mapped the other configuration. It will make it easier in the future if we replace the configurable-http-proxy with another implementation also that may merit from other defaults. The old name is deprecated with a `fail` to help users quickly spot the change and transition.

## Motivation for the new default

1. Most other Helm chart defaults to having PDBs disabled by default
   making our chart's behavior unexpected.

2. The idea of using a PDB is tightly coupled with applications that
   support high availability, in other words, to run in parallel to
   another. But, the hub/proxy/autohttps pod doesn't support it. As an
   indication of this, see the k8s docs on "configure-pdb" which has a
   "Before you begin" section that sais "You are the owner of an
   application running on a Kubernetes cluster that requires high
   availability.

   ref: https://kubernetes.io/docs/tasks/run-application/configure-pdb/

3. When evictions take place, such as when a node is drained, the PDBs
   will be respected, and since we have `minAvailable: 1` and cannot
   create another one, a manual pod deletion is required.

   It is my perception that the act of draining a node should come with
   intent and consideration typically. If it comes as part of an
   automated k8s upgrade of nodes, it should at least be scheduled at a
   acceptable time window etc. So, it wouldn't be helpful to add the
   need to run `kubect delete pod` alongside this.

4. If the wish is to avoid making the cluster autoscaler relocate a pod
   part of a deployment, then it is more direct and sensible to use the
   `"cluster-autoscaler.kubernetes.io/safe-to-evict": "false"`
   annotation.

   ref: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node

manics commented:
LGTM 😄 

Does schema.yml also need an update?
https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/dea8badd102ac6bd76e7c1df20b38afec83b0fb1/jupyterhub/schema.yaml#L889-L905

Also what do you think about copying your first commit message https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/1938/commits/7ccfd50a5ae275f8d3a38de7ad664e1b15e5bdb3
into the description of this PR? I think it's a good explanation, and it means if someone comes to the PR in future they'll see the rationale in the PR description without digging into the commits or reading through the linked issues.

consideRatio commented:
@manics thanks for your time reviewing this!!! :heart: :tada:

I copied the commit message over to the PR and updated schema.yaml by relocating the proxy.pdb config to proxy.chp.pdb as well as adding one for traefik. I refactored away the repeating text by using a YAML anchor as well.

manics commented on jupyterhub/schema.yaml:
```diff
@@ -562,10 +562,24 @@ properties:
-          Set the Pod Disruption Budget for the hub pod.
+          Configure a PodDisruptionBudget for this Deployment.
+
+          These are disabled by default for our deployments that doesn't support
```
```suggestion
          These are disabled by default for our deployments that don't support
```

manics commented on jupyterhub/schema.yaml:
```diff
@@ -562,10 +562,24 @@ properties:
+
+          These are disabled by default for our deployments that doesn't support
+          being run in parallel with multiple replicas. Only the user-scheduler
+          currently support being run in parallel with multiple replicas. If
```
```suggestion
          currently supports being run in parallel with multiple replicas. If
```

manics commented on jupyterhub/schema.yaml:
```diff
@@ -562,10 +562,24 @@ properties:
+
+          Note that if you aim to block scaling down a node with the
+          hub/proxy/autohttps pod that would cause disruptions of the
+          deployment, then you should instead annotate the pod's of the
```
```suggestion
          deployment, then you should instead annotate the pods of the
```

manics reviewed (commented):
A few typos but otherwise LGTM!

Off topic ... but since schema.yaml has come up, do you know anything about `values.schema.json`  https://www.arthurkoziel.com/validate-helm-chart-values-with-json-schemas/?

consideRatio commented:
:heart: your help with review matters a lot to me, thank you again!! :tada: :heart:

I find my focus bottlenecking if I have too many open active issues/prs, so I really appreciate helping moving them to a decisive outcome :)

consideRatio commented:
I've opened #1951 to represent the wish for better documentation about HA so that the we are more explicit with assumptions.

I suggest to merge this and not also do those docs update as part of this PR.

manics reviewed (approved):
I agree with merging this now. Worst case we can always revert this before release.

consideRatio commented:
:tada: 

I'll document this change as part of the changelog explicitly.

meeseeksmachine commented:
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/jupyterhub-helm-chart-0-11-0-released/7521/1
